### ISSUE TITLE
SparseFeatureStore take N-times more RAM than DenseFeatureStore

### ISSUE THREAD

**Issue by daxenberger** [dkpro/dkpro-tc#230]
Originally reported on Google Code with ID 230

```
Hi,

the SparseFeatureStore is extremely memory hungry if one of the LuceneNgram* features
is used.
My experiment uses for instance the LuceneCharacterNGramUFE feature.

If I use the DenseFeatureStore with this feature my max used RAM reaches ~200MB. If
I use the SparseFeatureStore it reaches 1.6 GB memory.
If I increase the amount of data in my experiment this easily exceeds the amount of
available machine memory.

I tested a bit and growth in memory demand seems to be caused by the LuceneFeatures.


Any suggestions what is the cause of this?

```

Reported by `Tobias.Horsmann` on 2014-12-23 16:12:21

**Comment by Horsmann** [comment]
The issue remains even after the various changes in the past.

It is also independent of using Lucene features or not
Here are two simple-test cases that print the ram usage, Sparse takes about 3x the memory

```
 @Test
    public void testitDense() throws TextClassificationException{
        featureStore = new DenseFeatureStore();
        for (int i=0; i < 99999; i++){
        Feature f1 = new Feature("feature1", "value1");
        Feature f2 = new Feature("feature2", "value2");
        Feature f3 = new Feature("feature3", "value3");
        Feature f4 = new Feature("feature4", "value4");
        Feature f5 = new Feature("feature5", "value5");
        List<Feature> features = new ArrayList<>();
        features.add(f1);
        features.add(f2);
        features.add(f3);
        features.add(f4);
        features.add(f5);
        Instance instance = new Instance(features, "outcome");
        featureStore.addInstance(instance);
        }

        System.out.println("DENSE: " + Runtime.getRuntime().totalMemory());
    }

    @Test
    public void testitSparse() throws TextClassificationException{
        featureStore = new SparseFeatureStore();
        for (int i=0; i < 999999; i++){
        Feature f1 = new Feature("feature1", "value1");
        Feature f2 = new Feature("feature2", "value2");
        Feature f3 = new Feature("feature3", "value3");
        Feature f4 = new Feature("feature4", "value4");
        Feature f5 = new Feature("feature5", "value5");
        List<Feature> features = new ArrayList<>();
        features.add(f1);
        features.add(f2);
        features.add(f3);
        features.add(f4);
        features.add(f5);
        Instance instance = new Instance(features, "outcome");
        featureStore.addInstance(instance);
        }

        System.out.println("SPARSE: " + Runtime.getRuntime().totalMemory());
    }
```

**Comment by reckart** [comment]
You'd probably save quite a bit of memory if you replace these collections

```
    private List<Map<String, Object>> instanceList = new ArrayList<>();
    private List<List<String>> outcomeList = new ArrayList<>();
    private List<Double> weightList = new ArrayList<>();
    private List<Integer> sequenceIds = new ArrayList<>();
    private List<Integer> sequencePositions = new ArrayList<>();
```

with equivalents from the [fastutil](http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22it.unimi.dsi%22%20AND%20a%3A%22fastutil%22) library.

**Comment by reckart** [comment]
E.g. `it.unimi.dsi.fastutil.ints.IntArrayList` instead of `ArrayList<Integer>`

**Comment by Horsmann** [comment]
Thanks I will take a look into it

**Comment by Horsmann** [comment]
Richards suggestion helped to ease the problem a little bit, but the problem with the memory consumption is that

`DenseFeatureStore` holds 2 data structures

```
    private List<Instance> instanceList;
    private TreeSet<String> featureNames;
```

`SparseFeatureStore` holds  5 data structures 

```
    private List<Map<String, Object>> instanceList = new ArrayList<>();
    private List<List<String>> outcomeList = new ArrayList<>();
    private DoubleArrayList weightList = new DoubleArrayList();
    private IntArrayList sequenceIds = new IntArrayList();
    private IntArrayList sequencePositions = new IntArrayList();
```

The essential difference is that the `SparseFeatureStore` does not hold the instance objects itself but rather their values and assembles on the fly a new `Instance` if requested while the `DenseFeatureStore` holds all `Instance`. 
To make this memory-friendly you have to reimplement it almost from scratch. This would probably only effect SvmHmm as all other modules use the `DenseFeatureStore`.

**Comment by reckart** [comment]
I think it should be possible to also replace these by fastutil classes:

```
    private List<Map<String, Object>> instanceList = new ArrayList<>();
    private List<List<String>> outcomeList = new ArrayList<>();
```

e.g. using `it.unimi.dsi.fastutil.objects.ObjectArrayList<K>`.

_Interning_ common strings may also help - in particular if the strings reoccur often (e.g. tags, lemmas, etc.)

If the elements of the outcomeList have a fixed size after creation, creating them as arrays should save space:

```
    private ObjectArrayList<String[]> outcomeList = new ObjectArrayList <>();
```

**Comment by Horsmann** [comment]
I suggest to not use the `SparseFeatureStore` unless you have a very good reason and a lot of RAM spare. I am not even really sure what it is suppose to do differently than then the `DenseFeatureStore` but what ever it does it needs a lot of data structures for. 
Are there objections to setting the `SparseFeatureStore` to deprecated?

**Comment by ferschke** [comment]
If you are dealing with many duplicate Strings, you might also want to think
a bit more about String management ­ e.g. Interning or deduplication

http://java-performance.info/java-string-deduplication
http://java-performance.info/string-intern-in-java-6-7-8/

**Comment by zesch** [comment]
I think that there is value in having a sparse feature store as e.g. ngram
features will in most cases be rather sparse. However, the current
implementation doesn't need to be the final word on that.

1) it is clear that the sparse feature store will use more memory if the
feature space is dense. So the memory test should actually account for this

2) in the dense feature store, we already account for string duplication by
storing feature names only once and only use numerical references in the
actual store

3) if you use context window features in sequence classification, the
resulting feature store will probably not be sparse, as this uses nominal
features. We have some other feature extractors that produce one-hot
representations, here we should see rather large savings.

To summarize:
- I would not deprecate the class now. First we need to make sure that we
  are testing it in sparse conditions.
- Even if we find that the current implementation is not optimal, there is
  value in improving it instead of abandoning it.

-Torsten

2016-05-13 16:16 GMT+02:00 Oliver Ferschke notifications@github.com:

**Comment by reckart** [comment]
> 2) in the dense feature store, we already account for string duplication by
> storing feature names only once and only use numerical references in the
> actual store

Interning strings is basically the Java convenience for what you describe. You only keep one copy of the string and store references to that same copy everywhere. It's just a bit more convenient than using actual numeric IDs.

**Comment by reckart** [comment]
You might want to consider storing token sequences as an array of interned strings rather than a new string, e.g.

```
["a", "rose", "is", "a", "rose", "is", "a", "rose"]
```

instead of

```
"a rose is a rose is a rose"
```

**Comment by Horsmann** [comment]
Maybe the issue are the feature extractors? The SparseFeatureStore takes only features which have a value set, right? The available features in TC are all fitted to the DenseStore i.e. always set some value that represents `not set` which is usually `not null`. 
Unless you use special features you cannot use the sparse feature store?

**Comment by Horsmann** [comment]
This issue is obsolete now. The sparse store was used with dense features which blew up the RAM consumptions. In issue #358 the features are now aware if the value is a `default value` i.e. if a feature is not set it is suppressed before it is handed to the `SparseFeatureStore`.
